### ISSUE TITLE
Rework how tracked structs etc are implemented for efficiency

### ISSUE THREAD

**Issue by nikomatsakis** [salsa-rs/salsa#312]
Currently, tracked structs, inputs and things are implemented "under the hood" as a distinct tracked function per field. This means a separate hashmap per field etc. This doesn't *feel* like the most efficient way. They don't really have interesting dependencies, for examples, and it requires a lot of locking to create a new tracked struct instance.

For tracked structs, it feels to me like we should store all the fields together, but keep a list of "last changed" revisions for each value field as well.

**Comment by zuobaoquan** [comment]
I did a naive benchmark that indicates creating tracked struct is way slow compared to vec-based arena.

```rust
struct Foo {
    a: usize,
    b: u64,
    c: u64,
    d: u64,
    e: u64,
}

#[salsa::tracked]
pub struct TrackedFoo {
    #[id]
    a: usize,
    b: u64,
    c: u64,
    d: u64,
    e: u64,
}
```

For 200_000 instances:

- tracked-new 382ms
- arena-push 3ms

**Comment by nikomatsakis** [comment]
I was going to say that a good first step is to have a benchmark. Just to set expectations, though, there is no way that we are going to be able to get a tracked struct as fast as pushing onto an area, though, especially with an `#[id]` field!

**Comment by zuobaoquan** [comment]
@nikomatsakis Yes. You are right. I just made simple tests to explore potential overhead  :-)

**Comment by nikomatsakis** [comment]
I was thinking about this question and I had an idea. Right now, when we store a tracked struct, we store an id. Accessing the fields or values off that struct requires locating the right ingredient and looking up the id in a (concurrent) hashmap. This is pretty expensive.

I was thinking about an alternative scheme where a tracked struct could be a pointer to an *actual struct* that stores the data for each field, along with some information about when they last changed. Those structs would, I think, have to be reference counted, so tracked structs could no longer be `Copy`. Cloning a handle to that struct would increment the reference count.

In the regular course of things, you would create a tracked struct. If the same struct existed in a prior revision, we would reuse the same memory, but overwrite the fields and update the versioning information. If not, we'd allocate a new chunk of memory. Just like today, we would ensure that the database retains a reference on the memory until incrementing the revision counter, allowing us to hand out references that are tied to the `&db`.

Reading a field could be done very cheaply, almost just a pointer read, except that we do want to check one invariant: when you read the field, we should check that the struct was "verified" in the current db revision. The idea is that the struct is verified either when it is created via `Struct::new` or when the creation query is itself verified. This guarantees that the values of the fields are up-to-date and will not change for the remainder of the revision, another thing which allows us to pass out an `&db` value. (In general, if you manage to read the field of a tracked struct before its been validated, you must be leaking the value or doing *something* bogus, so we should be checking for this anyway.)

We could make interning use a similar scheme, I believe.

I imagine this would all be quite a bit faster. At least the way I'm imagining it right now, though, it would mean that each tracked struct creation is an arc allocation. That's probably a lot cheaper than the current scheme, which winds up allocating a memo *per field*, but I do wonder if we could modify the scheme to use something more like an arena. I don't see why not, but the reference counts would have to be on the arena 'slabs' or something like that. (Because tracked struct types don't carry lifetimes, it's possible for the database to be freed and for there to still be tracked structs out there; so we need the ref counts to ensure that data they reference stays around.)

All of this would substantially increase the "unsafe code" footprint of salsa, but I think it may be worth it.

**Comment by nikomatsakis** [comment]
I think I will try to write up the docs for how this scheme will work in a PR first, before toying with the actual implementation. Part of my motivation for this is that this scheme is complex enough that I would want to ensure it's well documented.

**Comment by MichaReiser** [comment]
I think this is actually implemented today.
